packages/core/src/modules/connections/DidExchangeProtocol.ts

```diff
@@ -0,0 +1,462 @@
+
+    const { goal, goalCode, routing, autoAcceptConnection } = params
+    const label = params.label ?? this.config.label
+    const { verkey } = routing
+    const peerDid = await this.createPeerDidDoc(routing)
+    const parentThreadId = connectionRecord.invitation?.id
+
+    const message = new DidExchangeRequestMessage({ label, parentThreadId, did: peerDid.did, goal, goalCode })
+
+    // Create sign attachment containing didDoc
+    if (peerDid.numAlgo === PeerDidNumAlgo.GenesisDoc) {
+      const didDocAttach = await this.createSignedAttachment(peerDid.didDocument, verkey)
+      message.didDoc = didDocAttach
+    }
+
+    connectionRecord.did = peerDid.did
+    connectionRecord.verkey = verkey
+    connectionRecord.threadId = message.id
+
+    if (autoAcceptConnection !== undefined || autoAcceptConnection !== null) {
+      connectionRecord.autoAcceptConnection = autoAcceptConnection
+    }
+
+    await this.updateState(DidExchangeRequestMessage.type, connectionRecord)
+    this.logger.debug(`Create message ${DidExchangeRequestMessage.type} end`, {
+      connectionRecord,
+      message,
+    })
+    return message
+  }
+
+  public async processRequest(
+    messageContext: InboundMessageContext<DidExchangeRequestMessage>,
+    routing?: Routing
+  ): Promise<ConnectionRecord> {
+    this.logger.debug(`Process message ${DidExchangeRequestMessage.type} start`, messageContext)
+
+    // eslint-disable-next-line prefer-const
+    let { connection: connectionRecord, message } = messageContext
+
+    if (!connectionRecord) {
+      throw new AriesFrameworkError('No connection record in message context.')
+    }
+
+    DidExchangeStateMachine.assertProcessMessageState(DidExchangeRequestMessage.type, connectionRecord)
+
+    // check corresponding invitation ID is the request's ~thread.pthid
+
+    if (connectionRecord.invitation?.id !== message.thread?.parentThreadId) {
+      throw new ProblemReportError('Missing reference to invitation.', { problemCode: 'request_not_accepted' })
+    }
+
+    // If the responder wishes to continue the exchange, they will persist the received information in their wallet.
+
+    // Create new connection if using a multi use invitation
+    if (connectionRecord.multiUseInvitation) {
+      if (!routing) {
+        throw new AriesFrameworkError(
+          'Cannot process request for multi-use invitation without routing object. Make sure to call processRequest with the routing parameter provided.'
+        )
+      }
+
+      connectionRecord = await this.connectionService.createConnection({
+        role: connectionRecord.role,
+        state: connectionRecord.state,
+        multiUseInvitation: false,
+        routing,
+        autoAcceptConnection: connectionRecord.autoAcceptConnection,
+        invitation: connectionRecord.invitation,
+        tags: connectionRecord.getTags(),
+        protocol: connectionRecord.protocol,
+      })
+    }
+
+    const peerDid = DidPeer.fromDid(message.did)
```

Comment:
Maybe we should add a guard here that checks whether the did start with `did:peer:` and otherwise throw an error that public dids are not supported?

Reply:
It could be responsibility of `DidPeer` class. I added it anyway, because we would still need some error handling here so that check inside `DidPeer` wouldn't help so much.